### PR TITLE
Feat/server side filtering multiselect

### LINKED ISSUE #5018: [Feature] Enable Server Side Filtering on Dropdown Menu
It would be great if you could filter an Appsmith drop-down menu on the server side. I want to be able to select an item out of a list that will eventually include tens of thousands of options, and a drop-down is a much more natural way to do this than using a full table for selections (my current workaround.)

Ideally, the user would type in the drop-down's search box to filter the values and the application would send an updated DB query with the new filter, just like with paginated tables. I believe this would only require exposing the current value of the search box as a property, and firing an event with every change in it's value.

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

Add support for Server Side Filtering on MultiSelect

Fixes #5018 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

> Please describe the tests that you ran to verify your changes. Provide instructions, so we can reproduce.
> Please also list any relevant details for your test configuration.

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feat/Server-side-filtering-Multiselect 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.82 **(-0.02)** | 36.66 **(-0.02)** | 33.23 **(-0.03)** | 55.4 **(-0.02)**
 :green_circle: | app/client/src/components/designSystems/appsmith/MultiSelectComponent/index.tsx | 17.31 **(2.68)** | 0 **(0)** | 0 **(0)** | 18.75 **(2.96)**
 :red_circle: | app/client/src/components/designSystems/blueprint/DropdownComponent.tsx | 45.1 **(-0.73)** | 23.53 **(-3.14)** | 5.26 **(-0.3)** | 44.9 **(-0.75)**
 :red_circle: | app/client/src/widgets/DropdownWidget.tsx | 52.08 **(-4.74)** | 0 **(0)** | 45.45 **(-10.11)** | 55.81 **(-5.73)**
 :red_circle: | app/client/src/widgets/MultiSelectWidget.tsx | 50.98 **(-5.54)** | 0 **(0)** | 41.67 **(-8.33)** | 52.08 **(-6.06)**</details>

### CODE REVIEW

**Review by somangshu — CHANGES_REQUESTED** [review]
- ~~Order of the new props need to change~~
    - ~~`Server side filtering` after `disabled`~~
    - ~~`onFilterUpdate` the first under actions~~
- ~~`onFilterUpdate` should be a dropdown to bind actions which should called when filter is updated~~
- The data once reloaded should update. 
- ~~The `filterText` prop does not reset on blur whereas the searchText does.~~
- Same changes should be applied to the select widget too
- The action is not triggered on filterText update

**Comment by somangshu** [comment]
@Tooluloope ☝️

File: app/client/src/widgets/DropdownWidget.tsx [hunk 1]
```diff
@@ -80,6 +80,16 @@ class DropdownWidget extends BaseWidget<DropdownWidgetProps, WidgetState> {
             isTriggerProperty: false,
             validation: VALIDATION_TYPES.BOOLEAN,
           },
+          {
+            helpText: "Controls the Remote Data/ Server Side Filtering",
```
Comment (somangshu) [inline app/client/src/widgets/DropdownWidget.tsx:84]: Change to > "Enables server side filtering of the data"

File: app/client/src/components/designSystems/blueprint/DropdownComponent.tsx [hunk 2]
```diff
@@ -10,7 +10,7 @@ import {
 import { IconNames } from "@blueprintjs/icons";
 import { DropdownOption } from "widgets/DropdownWidget";
 import { Select, IItemRendererProps } from "@blueprintjs/select";
-import _ from "lodash";
+import _, { debounce } from "lodash";
```
Comment (somangshu) [inline app/client/src/components/designSystems/blueprint/DropdownComponent.tsx]: Why do we need to import `debounce` separately, And if there are no other references the `_` ref should be removed.

File: app/client/src/widgets/DropdownWidget.tsx [hunk 3]
```diff
@@ -94,6 +104,16 @@ class DropdownWidget extends BaseWidget<DropdownWidgetProps, WidgetState> {
             isBindProperty: true,
             isTriggerProperty: true,
           },
+          {
+            helpText: "Filter options using filterText",
```
Comment (somangshu) [inline app/client/src/widgets/DropdownWidget.tsx:108]: change to > Trigger an action on change of `filterText`

**Review by somangshu — CHANGES_REQUESTED** [review]

**Review by somangshu — CHANGES_REQUESTED** [review]
@Tooluloope please resolve conflict. Have some feedback from the deploy preview:
- The filterText in the multi-select widget does not call the onFilterUpdate action when the filter is completely remove (cmd + A and Backspace)
- The select widget does not have the filterText api exposed

**Comment by somangshu** [comment]
@Tooluloope please note that documentation changes are required here

**Comment by Tooluloope** [comment]
> @Tooluloope please resolve conflict. Have some feedback from the deploy preview:
> 
> * The filterText in the multi-select widget does not call the onFilterUpdate action when the filter is completely remove (cmd + A and Backspace)
> * The select widget does not have the filterText api exposed
> * error seen > Could not parse the binding `onFilterUpdate`

I realized there are some queries that wont work or throw some error if filterText is empty.
Also, I saw it as an extra network request that might not be used, so it only works when It has a value(filterText)
> The filterText in the multi-select widget does not call the onFilterUpdate action when the filter is completely remove (cmd + A and Backspace)

For the rest, I've fixed it.
@somangshu

**Comment by Tooluloope** [comment]
> @Tooluloope please note that documentation changes are required here

@somangshu I've created a PR for this on the docs repo, you could review that

**Comment by somangshu** [comment]
> > The filterText in the multi-select widget does not call the onFilterUpdate action when the filter is completely remove 
(cmd + A and Backspace)

We need a find a solution to this @Tooluloope, With this the widget will never we able to come to its initial state in case of server side filtering.

**Comment by Tooluloope** [comment]
> > > The filterText in the multi-select widget does not call the onFilterUpdate action when the filter is completely remove
> > > (cmd + A and Backspace)
> 
> We need a find a solution to this @Tooluloope, With this the widget will never we able to come to its initial state in case of server side filtering.

@somangshu  Fixed it

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Review by somangshu — DISMISSED** [review]
Running tests

**Comment by riodeuno** [comment]
@Tooluloope Let's get the multiselect widget merged into release, so that the diff in this PR will be smaller. That way it will be easier to review.

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Review by somangshu — DISMISSED** [review]

**Comment by riodeuno** [comment]
@somangshu Can we verify the UI/UX?

**Comment by somangshu** [comment]
Checked the deploy preview, Something is wrong. @Tooluloope the filter after completing the query run is reseting the select options, Also the server side filtering is overridden by the client side filter. 👇 

Uploading Screen Recording 2021-08-10 at 7.29.20 PM.mov…

**Review by somangshu — APPROVED** [review]
LGTM, Deploy preview also tested

**Review by riodeuno — APPROVED** [review]

**Pull request merged**